export/securedrop_export/disk/cli.py

```diff
@@ -1,429 +1,526 @@
+            pexpect.EOF,
+            pexpect.TIMEOUT,
+        ]  # type: PexpectList
+        unlock_error = Status.ERROR_UNLOCK_GENERIC
+
+        child = pexpect.spawn(command)
+        index = child.expect(prompt)
+        if index != 0:
+            logger.error("Did not receive disk unlock prompt")
+            raise ExportException(sdstatus=Status.ERROR_UNLOCK_GENERIC)
+        else:
+            logger.debug("Passing key")
+            child.sendline(encryption_key)
+            index = child.expect(expected)
+            if index == 0 or index == 1:
+                # Pexpect includes a re.Match object at `child.match`, but this freaks mypy out:
+                # see https://pexpect.readthedocs.io/en/stable/api/pexpect.html#pexpect.spawn.expect
+                # We know what format the results are in
+                dm_name = child.match.group(1).decode("utf-8").strip()  # type: ignore
+                logger.debug(f"Device is unlocked as {dm_name}")
+
+                child.close()
+                if child.exitstatus is not None and child.exitstatus not in (0, 1):
+                    logger.warning(f"pexpect: child exited with {child.exitstatus}")
+
+                # dm_name format is /dev/dm-X
+                return self._mount_volume(volume, dm_name)
+
+            elif index == 2:
+                # Still an error, but we can report more specific error to the user
+                logger.debug("Bad volume passphrase")
+                unlock_error = Status.ERROR_UNLOCK_LUKS
+
+            # Any other index values are also an error. Clean up and raise
+            child.close()
+
+            logger.error(f"Error encountered while unlocking {volume.device_name}")
+            raise ExportException(sdstatus=unlock_error)
+
+    def _mount_volume(self, volume: Volume, full_unlocked_name: str) -> MountedVolume:
         """
-        Given an unlocked LUKS volume, return MountedVolume object.
+        Given an unlocked volume, mount volume in /media/user/ by udisksctl and
+        return MountedVolume object.
 
-        If volume is already mounted, mountpoint is not changed. Otherwise,
-        volume is mounted at _DEFAULT_MOUNTPOINT.
+        Unlocked name could be `/dev/mapper/$id` or `/dev/dm-X`.
 
         Raise ExportException if errors are encountered during mounting.
+
+        `pexpect.ExeptionPexpect` can't be try/caught, since it's not a
+        child of BaseException, but instead, exceptions can be included
+        in the list of results to check for. (See
+        https://pexpect.readthedocs.io/en/stable/api/pexpect.html#pexpect.spawn.expect)
         """
-        if not volume.unlocked:
-            logger.error("Volume is not unlocked.")
-            raise ExportException(sdstatus=Status.ERROR_MOUNT)
 
-        mountpoint = self._get_mountpoint(volume)
+        info = f"udisksctl info --block-device {volume.device_name}"
```

Comment:
I didn't catch this in my original review.  We'd want more sanitization here if the parameters were anything less-trusted than device names.

Reply:
On further inspection, I suppose we *could* pass a list to [`pexpect.spawn()`](https://pexpect.readthedocs.io/en/stable/api/pexpect.html#pexpect.spawn) as we do to the `subprocess.*` methods.  I defer to you, @rocodes, on whether that's worthwhile.

Reply:
This is a good point/slightly interesting dilemma. Our approach hasn't changed since the initial sd-devices prototype: we accept and don't sanitize the results of lsblk (or udisks2), we don't send these strings back to more trusted domains (client), we do include them in debug logs. We should consider device names basically untrusted, but I think because they are in an untrusted domain and the journalist user has introduced them, we're ok to pass them to udisks2 in that domain. But if you have other thoughts I'm all ears.

Reply:
Somehow I didn't see your second comment til a few mins ago. 

I agree and updated pexpect commands to use a list in https://github.com/freedomofpress/securedrop-client/pull/1777/commits/dc32ba2f8cdc4412020eff3679349e11596e037a  , and did some basic testing with the cli module alone before pushing just for sanity's sake.  I also shell-escaped the device strings; it's not a perfect (udisks still supplies some of our device strings, and I think device strings should be considered untrusted generally), but we are also operating only in an untrusted domain, and the user has control over which devices they input, so it's a bit of a rabbithole.  At least now anywhere we pass a device string into a command it is escaped.

There's a similar issue open for the device encryption passphrase in the client. Right now that's not escaped, and is passed to pexpect. But the source is the journalist user and if they are typing shell commands into a trusted domain (sd-app) in order to mess with a less-trusted domain.... I think there are bigger issues :)  And I would like to look at that separately, since I think there's greater potential for breakage if I change that on the fly.